### PR TITLE
Update to have a conditional pipeline so we only run what is needed based on the pr

### PR DESCRIPTION
**Pull request by m-nash**
First step of https://github.com/microsoft/typespec/issues/3117

We will need to do the following after merge detailed in the above issue

-  update typespec - core - ci to point at the new `eng/common/pipelines/ci.yml`
- move tsp-core steps into a template inside `eng/tsp-core/pipelines/templates` so that these can be changed independently
- remove `eng/tsp-core/pipelines/ci.yml
- remove `packages/http-client-csharp/eng/pipelies/ci.yml`

### CODE REVIEW

File: eng/common/pipelines/ci.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
+trigger: none
+pr:
```
Comment (timotheeguerin) [inline eng/common/pipelines/ci.yml:2]: I assume this should be updated?

For typespec core I think also we would like to have this run on `main`

File: eng/common/pipelines/ci.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                displayName: Initialize
+
+      # Run csharp stages if RunCSharp == true
+      - template: /packages/http-client-csharp/eng/pipeline/templates/ci-stages.yml
+        parameters:
+          DependsOn: InitStage
+          Condition: eq('true', stageDependencies.InitStage.outputs['InitJob.InitStep.RunCSharp'])
+
+      # Run core stages if RunCore == true
+      - stage: Build_Core
```
Comment (timotheeguerin) [inline eng/common/pipelines/ci.yml:38]: should that just be a call to a template instead?

Comment (m-nash) [reply eng/common/pipelines/ci.yml:38]: yes, we are going to do that as a follow up step https://github.com/microsoft/typespec/issues/3117

File: eng/common/pipelines/ci.yml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        displayName: Build Core
```
Comment (timotheeguerin) [inline eng/common/pipelines/ci.yml:39]: nit but could we just name the stage to be the product so we reduce the text on the github status list

Comment (m-nash) [reply eng/common/pipelines/ci.yml:39]: It allows for multiple stages for a given product like this
image

Comment (timotheeguerin) [reply eng/common/pipelines/ci.yml:39]: hhm i see guess its ok then

**Review by timotheeguerin — APPROVED** [review]

File: eng/common/pipelines/ci.yml (hunk 3 above)
Comment (hallipr) [reply eng/common/pipelines/ci.yml:39]: We could reduce each product to a single stage, so there would just be:
```mermaid
flowchart LR
    Initialize --> Core & CSharp & Python
```

Comment (hallipr) [reply eng/common/pipelines/ci.yml:39]: Stages are just logical groupings for jobs. We should be able to combine them easily

**Pull request merged**